### ISSUE TITLE
Changing the core Point System (1.8-) to a logged Point System like Stackoverflow

### ISSUE THREAD

**Issue by q2apro** [q2a/question2answer#690]
I post this here as an **issue** because it is a severe issue as it causes confusion among users, regularly.

To understand how Stackoverflow can calculate points so accurately, I asked them: https://meta.stackoverflow.com/q/370194/1066234

They store votes and other point events per day. And they cannot be changed in the future!!

Contrary, the **"points recalculation"** of q2a is messing up all user points as soon as something has been changed, for instance: Points granted changed in the admin panel, existing posts get deleted (user points are reduced), new upvotes for old posts etc. – This is especially an issue for the monthly calculation of user points.

This is how Stackoverflow stores/tracks the reputation points:

```
total votes: 787
 2   9605154 (5)
-- 2012-03-07 rep +5    = 6         
 2  10531683 (5)
-- 2012-05-10 rep +5    = 11        
 1  11085525 (2)
-- 2012-06-18 rep +2    = 13        
 1  11178723 (2)
 2  11178599 (5)
-- 2012-06-24 rep +7    = 20        
 2  11178599 (5)
-- 2012-07-21 rep +5    = 25        
-- 2012-08-22 rep 0     = 25        
-- 2012-08-26 rep 0     = 25        
-- 2012-09-07 rep 0     = 25        
 1  12415046 (2)
 1  12420385 (2)
-- 2012-09-14 rep +4    = 29        
 2  12412388 (5)
-- 2012-09-15 rep +5    = 34      
...
```
This obviously means, they have **a separate table to store point events!**

You can check your own SO reputation (if you are member there): https://stackoverflow.com/reputation

And here is how they handle minus points:

> "Any deductions aren't made retrospectively" – ... if posts from the past are deleted, they are not changed in the past, but instead will trigger minuspoints for the current day!

For Q2A we need a core change to implement this. 

**Alternative: ** Maybe we write a monster plugin the overrides the entire point system. But I am 100 % sure that the stackoverflow system is the better one.

---

Another reason for a stacklike point storing: At the moment I have 19000 users, and about 500000 posts. The q2a recalculation of user points takes already now about 15 minutes! Even though I have a strong server at hand.

**Comment by svivian** [comment]
> They store votes and other point events per day. And they cannot be changed in the future!!

This isn't entirely true. When they reduced the points for question upvotes, they *did* recalculate historic points for all users ([see here](https://stackoverflow.blog/2010/03/19/important-reputation-rule-changes/)). I don't remember any other times they changed the number of points you get so I don't know if they ever did what you're suggesting.

As for logging points, we actually added a 'points' event for plugins a while back but it ended up causing performance problems, so it was removed. There are existing plugins that store daily reputation, and you could create a plugin yourself to log point changes (by hooking into the various events like `q_vote_up` and `q_hide`). Also don't forget that votes are now logged via the date columns in `qa_uservotes`.

One thing we could add in the admin page is a "Save" button that only saves the points options and doesn't recalculate. Besides that, I don't see a reason to add the rest to core.

**Comment by q2apro** [comment]
> Also don't forget that votes are now logged via the date columns in qa_uservotes.

Yes, there the upvotes, downvotes and flags are stored. However, not the best-answer-selects.

You might like to try the reputation history page from this plugin: https://github.com/q2apro/q2apro-userstatistics/blob/master/q2apro-reputation-page.php

I will think more about the general issue. Maybe I come up with another idea.

**Comment by q2apro** [comment]
I reached a "limit", the recalculation of 27600 users takes up to an hour ... we definitely have to implement the other system, so that recalculations are not necessary anymore!

**Comment by pupi1985** [comment]
I'm surprised you still consider this as an issue. As this has already been addressed by Scott above, I can't provide any additional information rather than asking questions.

1. **Why** did you have to run a recalculation of points?
2. **Why** did it have to run in less than an hour?
3. Have you considered the fact that, following your own example, recalculating points will take the same time it takes now **plus** the time it gets to insert 27600 records with all the point changes into that new table?

**Comment by q2apro** [comment]
> Why did you have to run a recalculation of points?

Good argument! I will track if there are any changes in the total scores. Maybe not. Then this would be completely unnecessary. I report back :-)

> Why did it have to run in less than an hour?

It is slowing down the whole site for this time. (Or only in my browser?!)

> recalculating points will take the same time it takes now plus the time it gets to insert 27600 records with all the point changes into that new table

As I said, the recalculation would be only once. When swapping to the new point system. Then in the future, it is not needed anymore.

**Comment by pupi1985** [comment]
1. Theoretically, you would only need to run a point recalculation if you make changes to any of the points in admin/points.
1. The process per se is heavy in database accesses. That means it might be slowing your website.
1. If the recalculation would only be needed **once**, then I might not be understanding how the point tracking you suggest is supposed to work. Suppose I have **once** run the recalculation you suggest. Then I modify the points for asking questions from 10 to 15. How would you calculate the amount of points that would be increased without performing some kind of recalculation plus inserting the new point changes?

**Comment by q2apro** [comment]
> 1. Theoretically, you would only need to run a point recalculation if you make changes to any of the points in admin/points.

True! But just yesterday I have realized why I did the recalculation once in a while, because I have 4 sites running and migrate posts from one to the other. Coincidentially a user reported yesterday that points are "not transferred", so I checked again and she was right. Now I am reindexing and readjusting the points after the migrate. So this should be the last thing where points were not regarded yet.

> Then I modify the points for asking questions from 10 to 15. How would you calculate the amount of points that would be increased without performing some kind of recalculation plus inserting the new point changes?
The new points would be regarded only from the day of change, the logging table would need to hold the points granted for that :( ... damn it. Stackoverflow is doing it like this, to my knowledge: https://meta.stackoverflow.com/q/370194/1066234

Seems that eventually I am really sticking with the recent way, since new problems might occur with the proposed point logging.

**Comment by mvhconsult** [comment]
A big downside of the current downvote system is that a user cannot view the downvotes received, just the total number leaving him/her wondering where the wrong/bad answer has been placed.
Creating a separate table with a link to which answer or question received the vote would help in this respect.
